File: MyLearning/DsaPractice/Graph/Islands.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace DsaPractice.Graph
{
    internal class Islands
    {
        /*
         * https://leetcode.com/problems/number-of-islands/
         * 
         *  Given an m x n 2D binary grid grid which represents a map of '1's (land) and '0's (water), return the number of islands.
            An island is surrounded by water and is formed by connecting adjacent lands horizontally or vertically.
            You may assume all four edges of the grid are all surrounded by water.
         *                     
         * Implicit Graph (Edges are move up, move down, move right, move left)
         *  -Am I given actual node connections?	✅ Explicit
            -Do I have to use rules (like movement) to find connections?	✅ Implicit
            -Is the input a grid, maze, or board?	✅ Implicit
         * 
         *                    0   1   2   3   4  
         *                  ----------------------
                          0 ["1","1","0","1","0"],
                          1 ["1","1","0","1","0"],
                          2 ["1","1","0","0","0"],
                          3 ["0","0","0","0","0"]

           BFS/DFS:  
            0 -> Connected: 0, 1, 2, 3
            0 -> (0, 1, 2, 3) All visited
            1 -> (1, 2, 3) All Visited
            2 -> (0, 1) All visited
            3 -> No connections
                    
         * Output: 1        
                              0   1   2   3   4
                            ---------------------
                          0 ["1","1","0","0","0"],
                          1 ["1","1","0","0","0"],
                          2 ["0","0","1","0","0"],
                          3 ["0","0","0","1","1"]

           BFS/DFS:  
            0 -> 
            0 -> Connected: 1
                    
          Output: 3
         * 
         * */

        public int NumIslands(char[][] grid)
        {
            int rows = grid.Length;
            int cols = grid[0].Length;

            int islands = 0;

            for (int row = 0; row < grid.Length; row++)
            {
                for (int col = 0; col < cols; col++)
                {
                    if (grid[row][col] == '1')
                    {
                        BFS(grid, row, col, rows, cols);
                        islands++;
                    }
                }
            }
            return islands;
        }

        private void BFS(char[][] grid, int row, int col, int rows, int cols)
        {
            Queue<(int row, int col)> queue = new();

            queue.Enqueue((row, col));
            grid[row][col] = '0';

            // top, down right, left
            int[] topDown = [-1, 1, 0, 0];
            int[] rightLeft = [0, 0, 1, -1];

            while (queue.Count > 0)
            {
                var current = queue.Dequeue();

                for (int i = 0; i < 4; i++)
                {
                    int newRow = current.row + topDown[i];
                    int newCol = current.col + rightLeft[i];

                    if (newRow >= 0 && newRow < rows && newCol >= 0 && newCol < cols)
                    {
                        if (grid[newRow][newCol] == '1' && grid[newRow][newCol] == '1')
                        {
                            queue.Enqueue((newRow, newCol));
                            grid[newRow][newCol] = '0';
                        }
                    }
                }
            }
        }

        public void Test()
        {
            char[][] grid = [
                ['1','1','0','0','0'],
                ['1','1','0','0','0'],
                ['0','0','1','0','0'],
                ['0','0','0','1','1']
            ];

            //int islands = NumIslands(grid);
            int islands = NumIslandsDFS(grid);
            Console.WriteLine(islands);
        }

        public int NumIslandsDFS(char[][] grid)
        {
            int rows = grid.Length;
            int cols = grid[0].Length;

            int islands = 0;

            for (int row = 0; row < grid.Length; row++)
            {
                for (int col = 0; col < cols; col++)
                {
                    if (grid[row][col] == '1')
                    {
                        DFS(grid, row, col, rows, cols);
                        islands++;
                    }
                }
            }
            return islands;
        }

        private void DFS(char[][] grid, int row, int col, int rows, int cols)
        {
            if (row < 0 || row >= rows || col < 0 || col >= cols || grid[row][col] == '0')
                return;

            grid[row][col] = '0'; // mark as visited 

            // top, down, right, left
            int[] topDown = [-1, 1, 0, 0];
            int[] rightLeft = [0, 0, 1, -1];

            for (int i = 0; i < 4; i++)
            {
                int newRow = row + topDown[i];
                int newCol = col + rightLeft[i];

                DFS(grid, newRow, newCol, rows, cols);
            }
        }
    }
}

File: MyLearning/PythonPractice/Graph/Islands.py
"""Question:
Count the number of islands in a binary grid (connected 1s).
"""

"""Notes:
Given an m x n 2D binary grid grid which represents a map of '1's (land) and '0's (water), return the number of islands.

An island is surrounded by water and is formed by connecting adjacent lands horizontally or vertically. You may assume all four edges of the grid are all surrounded by water.

Example 1:

Input: grid = [
  ["1","1","1","1","0"],
  ["1","1","0","1","0"],
  ["1","1","0","0","0"],
  ["0","0","0","0","0"]
]
Output: 1

Example 2:

Input: grid = [
  ["1","1","0","0","0"],
  ["1","1","0","0","0"],
  ["0","0","1","0","0"],
  ["0","0","0","1","1"]
]
Output: 3

 

Constraints:

    m == grid.length
    n == grid[i].length
    1 <= m, n <= 300
    grid[i][j] is '0' or '1'.


"""

from collections import deque
from typing import List, Tuple


from collections import deque

def num_islands_bfs(grid) -> int:
    rows, cols = len(grid), len(grid[0])
    islands = 0

    for r in range(rows):
        for c in range(cols):
            if grid[r][c] != "1":
                continue

            islands += 1
            q = deque([(r, c)])
            grid[r][c] = "0"

            while q:
                cr, cc = q.popleft()
                for dr, dc in [(-1,0),(1,0),(0,-1),(0,1)]:
                    nr, nc = cr + dr, cc + dc
                    if 0 <= nr < rows and 0 <= nc < cols and grid[nr][nc] == "1":
                        grid[nr][nc] = "0"
                        q.append((nr, nc))

    return islands



def num_islands_dfs(grid: List[List[str]]) -> int:
    # Same logic as BFS, but use recursion to flood-fill an island.
    rows, cols = len(grid), len(grid[0])

    def dfs(r: int, c: int) -> None:
        # Stop if out of bounds or not land.
        if r < 0 or r >= rows or c < 0 or c >= cols or grid[r][c] != "1":
            return
        # Mark this land cell as visited.
        grid[r][c] = "0"
        # Recurse in 4 directions.
        dfs(r - 1, c)
        dfs(r + 1, c)
        dfs(r, c - 1)
        dfs(r, c + 1)

    islands = 0
    for r in range(rows):
        for c in range(cols):
            # Each new unvisited land cell starts a new island traversal.
            if grid[r][c] == "1":
                islands += 1
                dfs(r, c)
    return islands
